File: javascript/036/main.js
/*
 * ProjectEuler 036
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v5.0.0
 * `node main.js`
 */

"use strict";

function main() {
    var sum = 0;
    for (var i = 1; i < 1000000; i++) {
        if (isPalindrome(i.toString(10)) && isPalindrome(i.toString(2))) {
            sum += i;
        }
    }
    console.log(sum);
}

function isPalindrome(str) {
    var len = str.length;
    for (var i = 0; i < len / 2; i++) {
        if (str[i] !== str[len - 1 - i]){
            return false;
        }
    }
    return true;
}

main();

File: haskell/036/Main.hs
{-
    ProjectEuler 036
    Haskell / Main.hs
    copyright (c) 2015 Susisu

    GHC 7.10.2
    `ghc Main.hs -o bin/main`
-}

module Main where

main :: IO ()
main =
    let f = (&&) <$> isPalindromic 2 <*> isPalindromic 10
    in print . sum $ filter f [1..1000000 - 1]

logIntBase :: (Integral a, Integral b, Integral c) => a -> b -> c
logIntBase base n = floor . logBase (fromIntegral base) $ fromIntegral n

isPalindromic :: Integral a => a -> a -> Bool
isPalindromic base n =
    let p  = logIntBase base n + 1
        ns = take p . map (`mod` base) $ iterate (`div` base) n
    in ns == reverse ns
